File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.py
class Solution:
    # Trick:
    # if sum(nums[:i]) % k == x and sum(nums[:j]) % k == x,
    # then sum(nums[i:j]) % k == 0
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        res = 0
        running_sum_mod = 0
        count = [1] + (k - 1) * [0]  # count the running sums per modulo

        for n in nums:
            running_sum_mod = (running_sum_mod + n) % k
            res += count[running_sum_mod]
            count[running_sum_mod] += 1

        return res

    # TLE
    """
    def subarraysDivByK(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [[None for _ in range(n)] for _ in range(n)]
        for i in range(n):
            dp[i][i] = nums[i] % k
        
        def recSearch(i: int, j: int) -> int:
            nonlocal dp
            if dp[i][j] is not None:
                return dp[i][j]
            if i == 0:
                dp[i][j] = recSearch(i, j-1) + nums[j]
            else:
                dp[i][j] = recSearch(i-1, j) - nums[i-1]
            dp[i][j] %= k
            return dp[i][j]

        res = 0
        for i in range(n):
            for j in range(i, n):
                if recSearch(i, j) == 0:
                    res += 1
        return res
    """

File: 0974-subarray-sums-divisible-by-k/0974-subarray-sums-divisible-by-k.rs
impl Solution {
    pub fn subarrays_div_by_k(nums: Vec<i32>, k: i32) -> i32 {
        let mut res: i32 = 0;
        let mut running_sum_mod: i32 = 0;
        let mut count: Vec<i32> = Vec::new();  // count the running sums per modulo
        count.push(1);
        for i in 0..(k-1) {
            count.push(0);
        }
        for n in &nums {
            running_sum_mod += *n;
            running_sum_mod = ((running_sum_mod % k) + k) % k;  // force value between 0 and k excluded
            res += count[running_sum_mod as usize];
            count[running_sum_mod as usize] += 1;
        }

        return res;
    }
}
